core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -487,7 +487,11 @@ protected void cleanAll() {
   }
 
   protected void deleteFile(String path) {
-    deleteFunc.accept(path);
+    try {
+      deleteFunc.accept(path);
+    } catch (RuntimeIOException ignored) {
+      // Allow other deletes to run even if this one fails
```

Comment:
Oddly enough this happened in the test and caused the TransactionTest to pass even with the bug. Because it failed to delete a file in `cleanAll` (probably a different issue worth investigating) it skipped part of the deletion causing the cached metadata to remain valid by chance.

Reply:
Found and fixed this bug as well (4th commit)